### ISSUE TITLE
Gauge controls (and others?) calls to UpdateLayout can cause re-entrant arrange pass which will crash

### ISSUE THREAD

**Issue by gardnerjr** [beto-rodriguez/Live-Charts#465]
#### How to reproduce?

I'm trying to create a trivial repro solution but I'll post the details first:

Create an app that has an `ObservableCollection` bound to a `ListBox`, with virtualization enabled and a `VirtualizingStackPanel` as the itemspanel
In the listbox, add an itemtemplate which contains a Gauge control bound to properties (like percent) in the items in the observable collection
Add a button that is bound to a command or handler that clears the listbox and re-adds items
Run the app, click the button.  See gauges.  Click the button again.
Possibly a crash, especially if there are language specific styles in your app.

The problem is something like:
* The observable collection is cleared.  This clears the underlying collection and fires a collection changed event
* the event causes the listbox to clear and rebuild its templated items
* causing the gauge charts to get ancestorchange and other binding related events as they are removed 
* causing them to call `UIElement. UpdateLayout` from inside `Gauge .Update`
* UpdateLayout causes an immediate layout pass of the entire tree containing the gauge control
* which will eventually cause the listbox which contained the gauges to try to measure/arrange 
* which will now throw an exception because the listbox's VirtualizingStackPanel hasn't gotten through the  collection change event yet and it tries to find/fix/whatever items that are no longer in the collection 
* that exception gets swallowed by binding update handlers, leaving the VSP in a bad state, which will  crash on the next call to UpdateLayout in the normal pass, hiding the original cause in a crash that just appears to be in VSP's arrange code with no other user code on the stack

i have a full callstack if you want, but a subset of it is below with some notes. It might be hard to get a full repro without a very complicated application that sets up language specific content/styles, etc.

I've worked around it temporarily by explicitly removing the items from the listbox that have gauges one by one first, then doing the `ObservableCollection.Clear` after those items are removed. 

#### Extra notes

It seems like you might be able to fix this just by removing the calls to `UpdateLayout` and replacing them with `InvalidateMeasure` so that they are rebuilt on the next layout pass, instead of forcing another layout pass from inside the individual controls.

See: 
https://msdn.microsoft.com/en-us/library/system.windows.uielement.updatelayout%28v=vs.110%29.aspx?f=255&MSPPError=-2147217396

> Therefore, you should avoid calling UpdateLayout after each incremental and minor change in the element tree. The layout system will perform element layout in a deferred manner, using an algorithm that balances performance and currency, and with a weighting strategy to defer changes to roots until all child elements are valid. You should only call UpdateLayout if you absolutely need updated sizes and positions, and only after you are certain that all changes to properties that you control and that may affect layout are completed.

#### I am using

<table>
<tr>
    <td>Live-Charts version</td>
    <td>0.7.x (but the calls to appear updatelayout still appear in master</td>
</tr>
<tr>
    <td>.Net Version</td>
    <td>4.6</td>
</tr>
<tr>
    <td>Windows</td>
    <td>10</td>
</tr>
</table>

### callstack that shows the problem 

*[NullReferenceException occurs here in  VSP, where its children have been started to be cleared but hasn't completed, so its child array has nulls in it because it is in the process of removing them when the layout pass occurs again]*
PresentationFramework.dll!System.Windows.Controls.VirtualizingStackPanel.ArrangeOverride(System.Windows.Size arrangeSize)
PresentationFramework.dll!System.Windows.FrameworkElement.ArrangeCore(System.Windows.Rect finalRect)
*[lots and lots of frames of measure/arrange]*
PresentationCore.dll!System.Windows.UIElement.Arrange(System.Windows.Rect finalRect)
PresentationFramework.dll!System.Windows.Controls.Control.ArrangeOverride(System.Windows.Size arrangeBounds)
PresentationFramework.dll!System.Windows.FrameworkElement.ArrangeCore(System.Windows.Rect finalRect)
PresentationCore.dll!System.Windows.UIElement.Arrange(System.Windows.Rect finalRect)
Microsoft.VisualStudio.Shell.14.0.dll!Microsoft.VisualStudio.PlatformUI.SplitterPanel.ArrangeOverride(System.Windows.Size finalSize)
PresentationFramework.dll!System.Windows.FrameworkElement.ArrangeCore(System.Windows.Rect finalRect)
PresentationCore.dll!System.Windows.UIElement.Arrange(System.Windows.Rect finalRect)
PresentationCore.dll!System.Windows.ContextLayoutManager.UpdateLayout()
PresentationCore.dll!System.Windows.UIElement.UpdateLayout()
*[here's the charts causing a full measure/arrange by directly calling update layout]*
LiveCharts.Wpf.dll!LiveCharts.Wpf.Gauge.Update()
LiveCharts.Wpf.dll!LiveCharts.Wpf.Gauge.UpdateCallback(System.Windows.DependencyObject dependencyObject, System.Windows.DependencyPropertyChangedEventArgs dependencyPropertyChangedEventArgs)
WindowsBase.dll!System.Windows.DependencyObject.OnPropertyChanged(System.Windows.DependencyPropertyChangedEventArgs e)
PresentationFramework.dll!System.Windows.FrameworkElement.OnPropertyChanged(System.Windows.DependencyPropertyChangedEventArgs e)
WindowsBase.dll!System.Windows.DependencyObject.NotifyPropertyChange(System.Windows.DependencyPropertyChangedEventArgs args)
WindowsBase.dll!System.Windows.DependencyObject.UpdateEffectiveValue(System.Windows.EntryIndex entryIndex, System.Windows.DependencyProperty dp, System.Windows.PropertyMetadata metadata, System.Windows.EffectiveValueEntry oldEntry, ref System.Windows.EffectiveValueEntry newEntry, bool coerceWithDeferredReference, bool coerceWithCurrentValue, System.Windows.OperationType operationType)
WindowsBase.dll!System.Windows.DependencyObject.InvalidateProperty(System.Windows.DependencyProperty dp, bool preserveCurrentValue)
PresentationFramework.dll!System.Windows.Data.BindingExpressionBase.Invalidate(bool isASubPropertyChange)
*[note here: transfer value catches exceptions, which actually eats this null exception.  however, things are now in a bad state and you'll get it on the next measure/arrange]*
PresentationFramework.dll!System.Windows.Data.BindingExpression.TransferValue(object newValue, bool isASubPropertyChange)
PresentationFramework.dll!System.Windows.Data.BindingExpression.HandlePropertyInvalidation(System.Windows.DependencyObject d, System.Windows.DependencyPropertyChangedEventArgs args)
PresentationFramework.dll!System.Windows.Data.BindingExpressionBase.OnPropertyInvalidation(System.Windows.DependencyObject d, System.Windows.DependencyPropertyChangedEventArgs args)
PresentationFramework.dll!System.Windows.Data.BindingExpression.OnPropertyInvalidation(System.Windows.DependencyObject d, System.Windows.DependencyPropertyChangedEventArgs args)
WindowsBase.dll!System.Windows.DependentList.InvalidateDependents(System.Windows.DependencyObject source, System.Windows.DependencyPropertyChangedEventArgs sourceArgs)
WindowsBase.dll!System.Windows.DependencyObject.NotifyPropertyChange(System.Windows.DependencyPropertyChangedEventArgs args)
*[lots and lots of stack frames inside wpf]*
PresentationFramework.dll!System.Windows.FrameworkElement.UpdateStyleProperty()
PresentationFramework.dll!System.Windows.TreeWalkHelper.InvalidateOnTreeChange(System.Windows.FrameworkElement fe, System.Windows.FrameworkContentElement fce, System.Windows.DependencyObject parent, bool isAddOperation)
PresentationFramework.dll!System.Windows.FrameworkElement.OnVisualParentChanged(System.Windows.DependencyObject oldParent)
PresentationFramework.dll!System.Windows.Controls.ListBoxItem.OnVisualParentChanged(System.Windows.DependencyObject oldParent)
PresentationCore.dll!System.Windows.Media.Visual.FireOnVisualParentChanged(System.Windows.DependencyObject oldParent)
PresentationCore.dll!System.Windows.Media.Visual.RemoveVisualChild(System.Windows.Media.Visual child)
PresentationCore.dll!System.Windows.Media.VisualCollection.DisconnectChild(int index)
PresentationCore.dll!System.Windows.Media.VisualCollection.Clear()
PresentationFramework.dll!System.Windows.Controls.UIElementCollection.ClearInternal()
PresentationFramework.dll!System.Windows.Controls.Panel.ClearChildren()
PresentationFramework.dll!System.Windows.Controls.Panel.EnsureEmptyChildren(System.Windows.FrameworkElement logicalParent)
PresentationFramework.dll!System.Windows.Controls.Panel.OnItemsChangedInternal(object sender, System.Windows.Controls.Primitives.ItemsChangedEventArgs args)
PresentationFramework.dll!System.Windows.Controls.VirtualizingPanel.OnItemsChangedInternal(object sender, System.Windows.Controls.Primitives.ItemsChangedEventArgs args)
PresentationFramework.dll!System.Windows.Controls.Panel.OnItemsChanged(object sender, System.Windows.Controls.Primitives.ItemsChangedEventArgs args)
PresentationFramework.dll!System.Windows.Controls.ItemContainerGenerator.OnRefresh()
PresentationFramework.dll!System.Windows.Controls.ItemContainerGenerator.OnCollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs args)
WindowsBase.dll!System.Windows.WeakEventManager.ListenerList<System.Collections.Specialized.NotifyCollectionChangedEventArgs>.DeliverEvent(object sender, System.EventArgs e, System.Type managerType)
WindowsBase.dll!System.Windows.WeakEventManager.DeliverEvent(object sender, System.EventArgs args)
WindowsBase.dll!System.Collections.Specialized.CollectionChangedEventManager.OnCollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs args)
PresentationFramework.dll!System.Windows.Data.CollectionView.OnCollectionChanged(System.Collections.Specialized.NotifyCollectionChangedEventArgs args)
PresentationFramework.dll!System.Windows.Controls.ItemCollection.OnViewCollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
WindowsBase.dll!System.Windows.WeakEventManager.ListenerList<System.Collections.Specialized.NotifyCollectionChangedEventArgs>.DeliverEvent(object sender, System.EventArgs e, System.Type managerType)
WindowsBase.dll!System.Windows.WeakEventManager.DeliverEvent(object sender, System.EventArgs args)
WindowsBase.dll!System.Collections.Specialized.CollectionChangedEventManager.OnCollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs args)
PresentationFramework.dll!System.Windows.Data.CollectionView.OnCollectionChanged(System.Collections.Specialized.NotifyCollectionChangedEventArgs args)
PresentationFramework.dll!System.Windows.Data.ListCollectionView.RefreshOverride()
PresentationFramework.dll!System.Windows.Data.CollectionView.RefreshInternal()
PresentationFramework.dll!System.Windows.Data.CollectionView.RefreshOrDefer()
PresentationFramework.dll!System.Windows.Data.ListCollectionView.ProcessCollectionChanged(System.Collections.Specialized.NotifyCollectionChangedEventArgs args)
PresentationFramework.dll!System.Windows.Data.CollectionView.OnCollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs args)
System.dll!System.Collections.ObjectModel.ObservableCollection<x>.OnCollectionChanged(System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
*[collection.clear is here at the bottom, causing the event to be fired and the problem to occur]*

**Comment by beto-rodriguez** [comment]
Thank you very much for this detailed report, I'll fix it

**Comment by idoan** [comment]
I tried what you offered, replacing UpdateLayout with InvalidateMeasure . although the error disappeared, this resulted in layout inconsistency, especially when my content's width changes ( e.g. from 1 digit to 2 digits).

another solution I tried is calling dispatch.invoke after InvlidateMeasure; this almost fixed the layout inconsistency but at the cost of seeing a delay when text changes from 1 digit to 2 digits.

I couldn't find any other reasonable solution for this.
